components/org.wso2.carbon.identity.application.authenticator.adapter/src/main/java/org/wso2/carbon/identity/application/authenticator/adapter/model/AuthenticationActionExecutionResult.java

```diff
@@ -0,0 +1,129 @@
+    private final Validity validity;
+    private final Availability availability;
+
+    public AuthenticationActionExecutionResult(String fieldName, Availability availability, Validity validity,
+                                               String message) {
+
+        this.fieldName = fieldName;
+        this.validity = validity;
+        this.message = message;
+        this.availability = availability;
+    }
+
+    public String getFieldPath() {
+
+        return FIELD_BASED_PATH + fieldName;
+    }
+
+    public Validity getValidity() {
+
+        return validity;
+    }
+
+    public String getMessage() {
+
+        return message;
+    }
+
+    public Availability getAvailability() {
+
+        return availability;
+    }
+
+    public String getFieldName() {
+
+        return fieldName;
+    }
+
+    public static void logSuccessResponseDataValidationError(
+            AuthenticationActionExecutionResult executionResult) {
+
+        logResponseExecutionResult(executionResult, DiagnosticLog.ResultStatus.FAILED,
+                String.format("An error occurred while handling the authentication action " +
+                        "response for the %s field.", executionResult.getFieldName()));
+    }
+
+    public static void logSuccessResponseWithDefaultsForMissingData(
+            AuthenticationActionExecutionResult executionResult) {
+
+        logResponseExecutionResult(executionResult, DiagnosticLog.ResultStatus.SUCCESS,
+                String.format("Since the %s field is missing from the authentication action " +
+                        "response, the default value is used.", executionResult.getFieldName()));
+    }
+
+    public static void logIncompleteResponseExecutionResult(
+            AuthenticationActionExecutionResult executionResult) {
+
+        logResponseExecutionResult(executionResult, DiagnosticLog.ResultStatus.FAILED,
+                "The authentication action response processing failed due to an invalid response " +
+                        "for the INCOMPLETE status.");
+    }
+
+    private static void logResponseExecutionResult(AuthenticationActionExecutionResult executionResult,
+                                                   DiagnosticLog.ResultStatus resultStatus, String message) {
+
+        if (LoggerUtils.isDiagnosticLogsEnabled()) {
+
+            DiagnosticLog.DiagnosticLogBuilder diagLogBuilder = new DiagnosticLog.DiagnosticLogBuilder(
+                    ActionExecutionLogConstants.ACTION_EXECUTION_COMPONENT_ID,
+                    ActionExecutionLogConstants.ActionIDs.PROCESS_ACTION_RESPONSE);
+            diagLogBuilder
+                    .inputParam(StringUtils.EMPTY, executionResult)
+                    .resultMessage(message)
+                    .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                    .resultStatus(resultStatus)
+                    .build();
+            LoggerUtils.triggerDiagnosticLogEvent(diagLogBuilder);
+        }
+    }
```

Comment:
Shall we move this methods to a different class as the Diagnostic logger. 
Seems this class is within the model, but actually this is something fully internal.

Diagnostic logger will have the operation result as an internal class

Reply:
we can do this refactoring as a second step as well